<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Mathew Chamberlain" />

<meta name="date" content="2021-02-25" />

<title>Mapping homologous gene symbols</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Mapping homologous gene symbols</h1>
<h4 class="author">Mathew Chamberlain</h4>
<h4 class="date">2021-02-25</h4>



<p>This vignette shows how to map homologous gene symbols from Macaca fascicularis to the human genome.</p>
<div id="load-the-essential-packages" class="section level1">
<h1>Load the essential packages</h1>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">require</span>(biomaRt)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="kw">require</span>(tidyverse)</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="kw">require</span>(SignacX)</span></code></pre></div>
<p>After mapping the reads to the Macaca fasccicularis genome, we load the genes, which were generated from the output of the <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">cellranger pipeline</a> from 10X Genomics.</p>
</div>
<div id="load-macaca-fascicularis-genes" class="section level1">
<h1>Load Macaca fascicularis genes</h1>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>features.tsv &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">&quot;fls/features.tsv.gz&quot;</span>, <span class="dt">header =</span> <span class="ot">FALSE</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="kw">head</span>(features.tsv)</span></code></pre></div>
<pre><code>##                   V1                 V2              V3
## 1 ENSMFAG00000044637              PGBD2 Gene Expression
## 2 ENSMFAG00000039056             ZNF692 Gene Expression
## 3 ENSMFAG00000030010             ZNF672 Gene Expression
## 4 ENSMFAG00000002737            SH3BP5L Gene Expression
## 5 ENSMFAG00000000508              LYPD8 Gene Expression
## 6 ENSMFAG00000040572 ENSMFAG00000040572 Gene Expression</code></pre>
</div>
<div id="map-homologous-genes" class="section level1">
<h1>Map homologous genes</h1>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="co"># get human and cyno gene symbols</span></span>
<span id="cb4-2"><a href="#cb4-2"></a>human.R95 &lt;-<span class="st"> </span><span class="kw">useMart</span>(<span class="dt">host =</span> <span class="st">&quot;jan2019.archive.ensembl.org&quot;</span>, <span class="dt">biomart =</span> <span class="st">&quot;ENSEMBL_MART_ENSEMBL&quot;</span>, <span class="dt">dataset =</span> <span class="st">&quot;hsapiens_gene_ensembl&quot;</span>)</span>
<span id="cb4-3"><a href="#cb4-3"></a>cyno.R95 &lt;-<span class="st"> </span><span class="kw">useMart</span>(<span class="dt">host =</span> <span class="st">&quot;jan2019.archive.ensembl.org&quot;</span>, <span class="dt">biomart =</span> <span class="st">&quot;ENSEMBL_MART_ENSEMBL&quot;</span>, <span class="dt">dataset =</span> <span class="st">&quot;mfascicularis_gene_ensembl&quot;</span>)</span>
<span id="cb4-4"><a href="#cb4-4"></a></span>
<span id="cb4-5"><a href="#cb4-5"></a><span class="co"># values = listeENSID: list of cynomolgus ensembl IDs to be retrieved.</span></span>
<span id="cb4-6"><a href="#cb4-6"></a>listeENSID =<span class="st"> </span>features.tsv<span class="op">$</span>V1</span>
<span id="cb4-7"><a href="#cb4-7"></a>orthologs &lt;-<span class="st"> </span><span class="kw">getLDS</span>(<span class="dt">attributes =</span> <span class="kw">c</span>(<span class="st">&quot;ensembl_gene_id&quot;</span>, <span class="st">&quot;external_gene_name&quot;</span>), <span class="dt">filters =</span> <span class="st">&quot;ensembl_gene_id&quot;</span>, </span>
<span id="cb4-8"><a href="#cb4-8"></a>    <span class="dt">values =</span> listeENSID, <span class="dt">mart =</span> cyno.R95, <span class="dt">attributesL =</span> <span class="kw">c</span>(<span class="st">&quot;hgnc_symbol&quot;</span>, <span class="st">&quot;ensembl_gene_id&quot;</span>), <span class="dt">martL =</span> human.R95)</span>
<span id="cb4-9"><a href="#cb4-9"></a>orthologs &lt;-<span class="st"> </span><span class="kw">as_tibble</span>(orthologs)</span>
<span id="cb4-10"><a href="#cb4-10"></a><span class="kw">colnames</span>(orthologs) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;GeneID&quot;</span>, <span class="st">&quot;cynoSymbol&quot;</span>, <span class="st">&quot;HumanSymbol&quot;</span>, <span class="st">&quot;HumanGeneID&quot;</span>)</span>
<span id="cb4-11"><a href="#cb4-11"></a></span>
<span id="cb4-12"><a href="#cb4-12"></a><span class="co"># keep only 1:1 orthologs</span></span>
<span id="cb4-13"><a href="#cb4-13"></a>one2one &lt;-<span class="st"> </span>orthologs <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">group_by</span>(GeneID) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">summarise</span>(<span class="kw">n</span>()) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(<span class="st">`</span><span class="dt">n()</span><span class="st">`</span> <span class="op">&lt;=</span><span class="st"> </span><span class="dv">1</span>) <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">select</span>(GeneID) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb4-14"><a href="#cb4-14"></a><span class="st">    </span><span class="kw">pull</span>()</span>
<span id="cb4-15"><a href="#cb4-15"></a>orthologs &lt;-<span class="st"> </span>orthologs <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(GeneID <span class="op">%in%</span><span class="st"> </span>one2one)</span>
<span id="cb4-16"><a href="#cb4-16"></a></span>
<span id="cb4-17"><a href="#cb4-17"></a><span class="co"># replace empty HumanSymbol (where there isn&#39;t a gene name for a homologous gene) with NA</span></span>
<span id="cb4-18"><a href="#cb4-18"></a>orthologs &lt;-<span class="st"> </span>orthologs <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">HumanSymbol =</span> <span class="kw">replace</span>(HumanSymbol, HumanSymbol <span class="op">==</span><span class="st"> &quot;&quot;</span>, <span class="ot">NA</span>))</span>
<span id="cb4-19"><a href="#cb4-19"></a>orthologs &lt;-<span class="st"> </span>orthologs <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">cynoSymbol =</span> <span class="kw">replace</span>(cynoSymbol, cynoSymbol <span class="op">==</span><span class="st"> &quot;&quot;</span>, <span class="ot">NA</span>))</span>
<span id="cb4-20"><a href="#cb4-20"></a></span>
<span id="cb4-21"><a href="#cb4-21"></a>idx =<span class="st"> </span><span class="kw">match</span>(listeENSID, orthologs<span class="op">$</span>GeneID)</span>
<span id="cb4-22"><a href="#cb4-22"></a>xx =<span class="st"> </span>orthologs<span class="op">$</span>HumanSymbol[idx]</span>
<span id="cb4-23"><a href="#cb4-23"></a>logik =<span class="st"> </span><span class="op">!</span><span class="kw">is.na</span>(orthologs<span class="op">$</span>HumanSymbol[idx])  <span class="co"># sum(logik) returns 17,365 homologous genes</span></span>
<span id="cb4-24"><a href="#cb4-24"></a>xx =<span class="st"> </span>xx[logik]</span>
<span id="cb4-25"><a href="#cb4-25"></a>orthologs =<span class="st"> </span>orthologs[<span class="op">!</span><span class="kw">is.na</span>(orthologs<span class="op">$</span>HumanSymbol), ]</span>
<span id="cb4-26"><a href="#cb4-26"></a><span class="co"># note: several of these genes are not unique mappers; we will aggregate them later or make them</span></span>
<span id="cb4-27"><a href="#cb4-27"></a><span class="co"># unique. To aggregate, where E is the sparse expression matrix with rownames set to xx: E =</span></span>
<span id="cb4-28"><a href="#cb4-28"></a><span class="co"># Matrix.utils::aggregate.Matrix(E, row.names(E))</span></span></code></pre></div>
<p>Now we have mapped homologous gene symbols across species:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="kw">head</span>(orthologs)</span></code></pre></div>
<pre><code>## # A tibble: 6 x 4
##   GeneID             cynoSymbol HumanSymbol HumanGeneID    
##   &lt;chr&gt;              &lt;chr&gt;      &lt;chr&gt;       &lt;chr&gt;          
## 1 ENSMFAG00000046426 ND6        MT-ND6      ENSG00000198695
## 2 ENSMFAG00000002805 POLG2      POLG2       ENSG00000256525
## 3 ENSMFAG00000046418 COX2       MT-CO2      ENSG00000198712
## 4 ENSMFAG00000042657 SLC38A3    SLC38A3     ENSG00000188338
## 5 ENSMFAG00000042891 HMOX1      HMOX1       ENSG00000100292
## 6 ENSMFAG00000038079 SHISA9     SHISA9      ENSG00000237515</code></pre>
<p>After mapping homologous genes, Signac can be used to classify the cell types.</p>
<details>
<p><summary><strong>Session Info</strong></summary></p>
<pre><code>## R version 3.5.0 (2018-04-23)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: CentOS Linux 7 (Core)
## 
## Matrix products: default
## BLAS: /usr/lib64/libblas.so.3.4.2
## LAPACK: /site/ne/app/x86_64/R/v3.5.0/lib64/R/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=C              
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] SignacX_2.2.0
## 
## loaded via a namespace (and not attached):
##   [1] backports_1.1.6      plyr_1.8.6           igraph_1.2.5        
##   [4] lazyeval_0.2.2       splines_3.5.0        listenv_0.8.0       
##   [7] usethis_1.6.1        ggplot2_3.3.3        digest_0.6.18       
##  [10] htmltools_0.4.0      gdata_2.18.0         fansi_0.4.1         
##  [13] magrittr_1.5         memoise_1.1.0        tensor_1.5          
##  [16] neuralnet_1.44.2     cluster_2.1.0        ROCR_1.0-7          
##  [19] remotes_2.2.0        globals_0.12.5       matrixStats_0.56.0  
##  [22] prettyunits_1.1.1    colorspace_1.4-0     rappdirs_0.3.1      
##  [25] ggrepel_0.9.1        xfun_0.12            dplyr_0.8.5         
##  [28] callr_3.4.3          crayon_1.3.4         jsonlite_1.6.1      
##  [31] roxygen2_7.1.0       lme4_1.1-23          spatstat_1.64-1     
##  [34] spatstat.data_2.0-0  survival_3.2-7       zoo_1.8-4           
##  [37] ape_5.3              glue_1.4.0           polyclip_1.10-0     
##  [40] gtable_0.3.0         leiden_0.3.3         pkgbuild_1.0.6      
##  [43] future.apply_1.4.0   abind_1.4-5          scales_1.1.0        
##  [46] miniUI_0.1.1.1       Rcpp_1.0.4.6         viridisLite_0.3.0   
##  [49] xtable_1.8-4         reticulate_1.18      rsvd_1.0.3          
##  [52] htmlwidgets_1.5.3    httr_1.4.2           gplots_3.0.3        
##  [55] RColorBrewer_1.1-2   ellipsis_0.3.0       Seurat_3.2.0        
##  [58] ica_1.0-2            pkgconfig_2.0.3      uwot_0.1.8          
##  [61] deldir_0.2-10        utf8_1.1.4           RJSONIO_1.3-1.4     
##  [64] tidyselect_1.1.0     rlang_0.4.8          reshape2_1.4.4      
##  [67] later_1.0.0          munsell_0.5.0        pbmcapply_1.5.0     
##  [70] tools_3.5.0          cli_2.0.2            devtools_2.3.1      
##  [73] ggridges_0.5.3       evaluate_0.14        stringr_1.4.0       
##  [76] fastmap_1.0.1        yaml_2.2.1           goftest_1.2-2       
##  [79] npsurv_0.4-0         processx_3.4.2       knitr_1.28          
##  [82] fs_1.4.1             fitdistrplus_1.0-14  caTools_1.17.1.1    
##  [85] purrr_0.3.3          RANN_2.6.1           pbapply_1.4-2       
##  [88] future_1.16.0        nlme_3.1-145         mime_0.9            
##  [91] formatR_1.7          xml2_1.3.1           compiler_3.5.0      
##  [94] rstudioapi_0.11      plotly_4.9.3         png_0.1-7           
##  [97] testthat_2.3.2       lsei_1.2-0           spatstat.utils_2.0-0
## [100] tibble_3.0.6         statmod_1.4.34       stringi_1.4.6       
## [103] ps_1.3.2             desc_1.2.0           lattice_0.20-41     
## [106] Matrix_1.2-18        nloptr_1.2.2.1       vctrs_0.3.6         
## [109] pillar_1.4.3         lifecycle_0.2.0      lmtest_0.9-37       
## [112] RcppAnnoy_0.0.16     data.table_1.14.0    cowplot_1.1.1       
## [115] bitops_1.0-6         irlba_2.3.3          httpuv_1.5.2        
## [118] patchwork_1.1.1      R6_2.4.1             promises_1.1.0      
## [121] KernSmooth_2.23-16   gridExtra_2.3        sessioninfo_1.1.1   
## [124] codetools_0.2-16     boot_1.3-24          MASS_7.3-51.1       
## [127] gtools_3.8.1         assertthat_0.2.1     pkgload_1.0.2       
## [130] rprojroot_1.3-2      withr_2.4.1          sctransform_0.3.2   
## [133] mgcv_1.8-31          parallel_3.5.0       grid_3.5.0          
## [136] rpart_4.1-15         tidyr_1.1.2          minqa_1.2.4         
## [139] rmarkdown_2.1        Rtsne_0.15           shiny_1.4.0.2</code></pre>
</details>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
